src/flamegraph/color.rs

```diff
@@ -0,0 +1,272 @@
+        Palette::Yellow
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_perl_palette(s: &str) -> Palette {
+    if s.contains("::") {
+        Palette::Yellow
+    } else if s.contains("Perl") || s.contains(".pl") {
+        Palette::Green
+    } else if s.ends_with("_[k]") {
+        Palette::Orange
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_js_palette(s: &str) -> Palette {
+    if s.ends_with("_[j]") {
+        if s.contains("/") {
+            return Palette::Green
+        } else {
+            return Palette::Aqua
+        }
+    } else if s.contains("::") {
+        return Palette::Yellow
+    } else if let Some(ai) = s.find("/") {
+        if (&s[ai..]).contains(".js") {
+            return Palette::Green
+        }
+    }
+
+    if s.contains(":") {
+        Palette::Aqua
+    } else if s == " " {
+        Palette::Green
+    } else if s.contains("_[k]") {
+        Palette::Orange
+    } else {
+        Palette::Red
+    }
+}
+
+fn handle_wakeup_palette(_s: &str) -> Palette {
+    Palette::Aqua
+}
+
+fn handle_chain_palette(s: &str) -> Palette {
+    if s.contains("_[w]") {
+        Palette::Aqua
+    } else {
+        Palette::Blue
+    }
+}
+
+fn coefficients_for_palette(palette: &Palette, name: &str, v1: f32, v2: f32, v3: f32) ->
+    ((u8, u8, f32), (u8, u8, f32), (u8, u8, f32)) {
+    let real_palette = match palette {
+        Palette::Hot => return ((205, 50, v3), (0, 230, v1), (0, 55, v2)),
+        Palette::Mem => return ((0, 0, v3), (190, 50, v2), (0, 210, v1)),
+        Palette::Io => return ((80, 60, v1), (80, 60, v1), (190, 55, v2)),
+        Palette::Red => return ((200, 55, v1), (50, 80, v1), (50, 80, v1)),
+        Palette::Green => return ((50, 60, v1), (200, 55, v1), (50, 60, v1)),
+        Palette::Blue => return ((80, 60, v1), (80, 60, v1), (205, 60, v1)),
+        Palette::Yellow => return ((175, 55, v1), (175, 55, v1), (50, 20, v1)),
+        Palette::Purple => return ((190, 65, v1), (80, 60, v1), (190, 65, v1)),
+        Palette::Aqua => return ((50, 60, v1), (165, 55, v1), (165, 55, v1)),
+        Palette::Orange => return ((190, 65, v1), (90, 65, v1), (0, 0, v1)),
+        Palette::Java => handle_java_palette(name),
+        Palette::Perl => handle_perl_palette(name),
+        Palette::Js => handle_js_palette(name),
+        Palette::Wakeup => handle_wakeup_palette(name),
+        Palette::Chain => handle_chain_palette(name),
+    };
+
+    coefficients_for_palette(&real_palette, name, v1, v2, v3)
+}
+
+fn affine_transform(a: u8, b: u8, x: f32) -> u8 {
+    b + (a as f32 * x) as u8
+}
+
+fn color_from_palette(palette: &Palette, name: &str, v1: f32, v2: f32, v3: f32) -> String {
+    let ((r_b, r_a, r_x),
+         (g_b, g_a, g_x),
+         (b_b, b_a, b_x)) = coefficients_for_palette(palette, name, v1, v2, v3);
+
+    let red = affine_transform(r_a, r_b, r_x);
+    let green = affine_transform(g_a, g_b, g_x);
+    let blue = affine_transform(b_a, b_b, b_x);
```

Comment:
I wonder if we should just do:
```rust
macro_rules! t {
  ($a:expr, $b:expr, $x:expr) => $b + ($a as f32 * $x) as u8
}
```
And then use `t!` directly inline in `coefficients_for_palette`. That would obviate the need for the extra destructuring and values here.

Reply:
[Done](https://github.com/jonhoo/inferno/pull/32/commits/d5000681e16a220684e72429b32d5115588c483a)